### PR TITLE
fix(storage): persist typed raw failure authority (#3897)

### PR DESCRIPTION
**Pull request by Sinity**
## Summary

Close the reviewed raw-failure authority gaps at exact head `a74e948a0ba5b77a34581c40ec6b3e7616d2da5e`, rebased onto current `origin/master` including #3902. The implementation preserves retained raw bytes, provider-neutral CAS evidence, legacy selectors, and fail-closed replay semantics.

## Problem

The exact-head review found that successful CAS resolution could remain lifecycle-terminal, validation-failed evidence lacked a structural trusted disposition boundary, repair candidate selection could bypass validation authority, and a later parser failure could reuse an earlier failure carrier. Legacy migration fixtures also retained schema objects that did not exist at their declared historical versions.

## Solution

- Keep CAS resolution receipts outside the terminal/deferred failure lifecycle and retire stale authority before an untyped current failure or a new CAS frontier attempt.
- Require trusted worker-disposition provenance for validation-failed corrupt-input explanations.
- Keep replay selection bound to exact raw coordinates, valid typed lifecycle evidence, and provider-neutral CAS semantics.
- Preserve strict unknown-provider decoding, typed worker diagnostics, terminal status projection, legacy selectors, and forced-reparse behavior.
- Make v7, v13, v14, v20, and v24 migration fixtures accurately omit later receipt/index objects so historical copy-forward tests exercise the real migration path, including restoring the pre-v30 raw-artifact index shape.

## Acceptance criteria

| Criterion | Disposition | Evidence |
| --- | --- | --- |
| Async success supersedes stale deferred CAS authority atomically | Satisfied | `test_persist_batch_success_supersedes_deferred_cas_evidence_in_source_transaction` |
| Corrupt-input outcomes remain terminal and explainable | Satisfied | `test_persist_batch_corrupt_input_remains_terminal_in_lifecycle`; public-route readiness regression |
| Validation-failed evidence cannot authorize unsupported or deferred carriers | Satisfied | `test_validation_failed_unsupported_terminal_evidence_remains_unexplained` and trusted-provenance validation |
| Unknown mixed JSONL cannot be partially accepted through prefetch | Satisfied | `test_unknown_mixed_jsonl_prefetch_falls_back_to_strict_decode` |
| Terminal unsupported shapes stop ordinary retry but remain force-reparseable | Satisfied | `test_parse_backlog_excludes_terminal_failure_authority_until_forced_reparse` |
| Malformed terminal evidence cannot suppress retry | Satisfied | `test_parse_backlog_keeps_malformed_terminal_evidence_retryable` |
| Generic failures cannot reuse stale authority and historical migration fixtures remain valid | Satisfied | `test_generic_parse_state_failure_retires_prior_failure_authority`; 7 focused migration tests passed |
| Live census, deployment, backup-gated re-observation, reviewed apply, and production reprocessing | Partial | Remains under `polylogue-reindex-source-remediation`; no live mutation was performed |

## Bead disposition

| Bead | Disposition | Remaining scope |
| --- | --- | --- |
| `polylogue-dyica` | Partial | Current production census, deployment, fresh backup authorization, historical raw re-observation or migration, reviewed dry-run and apply, immutable receipts, and live reprocessing remain under `polylogue-reindex-source-remediation`. |

## Verification

```text
focused managed raw-failure routes: 33 passed in 16.54s
focused authority regressions: 3 passed in 5.17s
historical migration fixture routes: 7 passed in 5.54s
devtools verify --quick at `a74e948a0`: 24 steps passed, exit 0
devtools verify layering --json: {"violations": [], "count": 0, "baselined_count": 311}
git diff --check: passed
```

The full test suite was not run. A prior broader exploratory selection reported 150 passed and 14 failures; the migration-fixture failures were subsequently repaired, and the remaining identified failures were unrelated repair paths blocked by unreleased durable change trains for source versions 27 through 30. No Beads command, live production mutation, live migration, production reprocessing, or merge was run.

## Review disposition

The exact-head review findings are covered by the preceding repair commits and `f2c3370b2` / `0c0e92807` / `a74e948a0`. Adversarial iteration 1 found that historical fixtures preinstalled v30 raw-artifact indexes; `a74e948a0` restores the pre-v30 shape and the seven migration routes pass. The PR remains open, non-draft, and unmerged.

### CODE REVIEW

**Comment by Sinity** [comment]
@codex review

**Comment by Sinity** [comment]
@codex review

**Comment by Sinity** [comment]
@codex review

**Comment by Sinity** [comment]
@codex review

**Comment by Sinity** [comment]
@codex review

**Comment by Sinity** [comment]
@codex review

File: polylogue/pipeline/services/ingest_batch/_core.py [hunk 1]
```diff
@@ -1181,6 +1181,11 @@ def _record_outcome(summary: _IngestBatchSummary, ir: IngestRecordResult) -> Non
         parse_error=ir.parse_error,
         error=ir.error,
         had_sessions=bool(ir.sessions),
+        outcome_code=ir.outcome_code,
+        retryable=ir.retryable,
+        evidence_ref=ir.evidence_ref,
+        remediation=ir.remediation,
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/pipeline/services/ingest_batch/_core.py:1188]: Implemented in 038eee4f4. The ordinary `process_ingest_batch` route now maps terminal worker dispositions to a validated closed raw-failure carrier and persists the typed outcome code, retryability, evidence reference, remediation, diagnostic, and exact raw coordinate through the source `bulk_connection()` transaction. The production-route regressions prove retained-raw identity, closed kind/status, lifecycle readiness/status projection, and rollback atomicity: `test_persist_batch_raw_state_updates_persists_terminal_worker_disposition_to_source` and `test_persist_batch_raw_state_updates_rolls_back_typed_evidence_with_raw_state`.

File: polylogue/storage/sqlite/migrations/source/030_raw_failure_coordinate_carriers.sql [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    'terminal_corrupt_input',
+    'terminal_superseded_deferred_cas_frontier',
+    'terminal_unknown_json_decode',
+    'terminal_unknown_export_no_session',
+    'terminal_unsupported_shape'
+);
+
+CREATE UNIQUE INDEX IF NOT EXISTS idx_raw_artifacts_failure_identity
+ON raw_artifacts(raw_id, origin, source_path, source_index)
+WHERE artifact_kind IN (
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/storage/sqlite/migrations/source/030_raw_failure_coordinate_carriers.sql:24]: Implemented in 038eee4f4. `read_raw_failure_lifecycle` now filters recency and sampling to valid closed raw-failure evidence pairs before selecting the latest carrier, so a newer ordinary artifact cannot mask typed evidence. Ordinary artifacts remain stored at the coordinate. Regression: `test_failed_raw_lifecycle_ignores_newer_ordinary_artifact_at_same_coordinate`; the focused raw-failure suite passed 17 tests and `devtools verify --quick` passed all 24 steps.

**Comment by Sinity** [comment]
@codex review

**Comment by Sinity** [comment]
Scope for the next repair commit(s): address the four exact-head lifecycle gaps identified in comments 3742338895, 3742338898, and 3742338901. I will make async batch success supersede stale deferred CAS authority with the same transactional semantics as the synchronous route; align typed terminal corrupt-input evidence with validation/readiness lifecycle rules and cover zero-length/decode persistence; carry strict unknown-provider JSONL decoding through prefetch and prevent malformed mixed input from being partially accepted; and exclude terminal unsupported-shape rows from parse backlog retry. I will preserve the source-v30 coordinate-carrier migration and backup requirement, add real-route regressions, run focused tests plus one devtools verify and one devtools verify --quick at the publish boundary, then update the exact-head PR carrier/body. No Beads mutation, live migration, production reprocessing, or merge is in scope.

File: polylogue/storage/repair.py [hunk 3]
```diff
@@ -3928,22 +3920,11 @@ def _raw_materialization_candidate_ids(
...
-                )
-                OR r.parse_error IN (
-                  '{_LEGACY_OLDER_ACCEPTED_FRONTIER_ERROR}',
-                  '{_LEGACY_UNCONVERTIBLE_BYTE_HEAD_ERROR}'
+                OR EXISTS (
+                  SELECT 1
+                  FROM raw_artifacts AS retry_evidence
+                  WHERE retry_evidence.raw_id = r.raw_id
+                    AND retry_evidence.artifact_kind IN ({", ".join("?" for _ in RAW_FAILURE_DEFERRED_EVIDENCE_KINDS)})
                 )
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/storage/repair.py:3962]: Preserved. The legacy exception-prefix and parse-error compatibility predicates remain in the historical repair selector until the backup-gated migration or re-observation owned by the successor scope.

File: polylogue/storage/sqlite/archive_tiers/revision_governance.py [hunk 4]
```diff
@@ -3035,6 +3028,25 @@ def mark_raw_parse_failed(
     store: RawRevisionGovernanceHost, raw_id: str, *, provider: Provider, error: BaseException
 ) -> None:
     """Persist a bounded parse/index failure for retained raw evidence."""
+    if provider is Provider.CODEX and isinstance(error, RawCASFrontierError):
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/storage/sqlite/archive_tiers/revision_governance.py]: Preserved. Membership replay records provider-neutral deferred CAS evidence for every provider that reaches the governance path, with the non-Codex route covered by the existing focused regression.

File: polylogue/storage/repair.py [hunk 5]
```diff
@@ -3928,22 +3920,11 @@ def _raw_materialization_candidate_ids(
...
-                  r.parse_error LIKE '{_MEMBERSHIP_REPLAY_CONFLICT_ERROR_PREFIX}%'
-                )
-                OR r.parse_error IN (
-                  '{_LEGACY_OLDER_ACCEPTED_FRONTIER_ERROR}',
-                  '{_LEGACY_UNCONVERTIBLE_BYTE_HEAD_ERROR}'
+                OR EXISTS (
+                  SELECT 1
+                  FROM raw_artifacts AS retry_evidence
+                  WHERE retry_evidence.raw_id = r.raw_id
+                    AND retry_evidence.artifact_kind IN ({", ".join("?" for _ in RAW_FAILURE_DEFERRED_EVIDENCE_KINDS)})
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/storage/repair.py]: Addressed in the branch. Retry authority now matches the retained raw origin, source path, source index, and typed support pair; the malformed-coordinate backlog regression proves a mismatched carrier remains retryable.

File: polylogue/daemon/status.py [hunk 6]
```diff
@@ -955,8 +978,11 @@ def _archive_raw_failure_info(
                 val_status = str(row[3] or "") if row[3] else ""
                 val_err = str(row[4] or "") if row[4] else ""
                 origin = str(row[1]) if row[1] else None
-                if "JSONDecodeError" in parse_err or "decode error" in parse_err.lower():
-                    kind: Literal["decode_error", "parse_error", "schema_violation", "unknown"] = "decode_error"
+                artifact_kind = str(row[5]) if row[5] is not None else None
+                if artifact_kind in RAW_FAILURE_EVIDENCE_KINDS:
+                    kind = cast(Any, artifact_kind)
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/daemon/status.py]: Addressed by the existing lifecycle-backed status projection. Status consumes validated lifecycle evidence rather than presenting an invalid kind/support pair as authoritative.

File: polylogue/sources/live/batch.py [hunk 7]
```diff
@@ -2560,7 +2582,11 @@ def _ingest_full_records_archive(
                             source_path=record.source_path,
                             source_index=record.source_index or 0,
                             acquired_at_ms=acquired_at_ms,
-                            kind=RawFailureEvidenceKind.TERMINAL_UNSUPPORTED_SHAPE,
+                            kind=(
+                                RawFailureEvidenceKind.TERMINAL_UNKNOWN_EXPORT_NO_SESSION
+                                if provider is Provider.UNKNOWN
+                                else RawFailureEvidenceKind.TERMINAL_UNSUPPORTED_SHAPE
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/sources/live/batch.py:2618]: Preserved. Unknown malformed JSONL is classified through the strict decode route, and the current mixed-record prefetch regression proves the terminal decode carrier is retained.

File: polylogue/core/raw_failure_evidence.py [hunk 8]
```diff
@@ -16,35 +16,69 @@ class RawFailureEvidenceKind(StrEnum):
...
     DEFERRED_HOT_JSONL_CAPTURE = "deferred_hot_jsonl_capture"
+    DEFERRED_CLAUDE_CODE_PARTIAL_JSONL = "deferred_claude_code_partial_jsonl"
+    DEFERRED_CAS_FRONTIER = "deferred_cas_frontier"
+    # Historical rows written before CAS evidence was made provider-neutral.
+    # Keep this token readable until a backup-gated migration or re-observation
+    # receipt has converted every retained row.
+    DEFERRED_CODEX_CAS_FRONTIER = "deferred_codex_cas_frontier"
     TERMINAL_CORRUPT_INPUT = "terminal_corrupt_input"
+    TERMINAL_UNKNOWN_JSON_DECODE = "terminal_unknown_json_decode"
+    TERMINAL_UNKNOWN_EXPORT_NO_SESSION = "terminal_unknown_export_no_session"
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/core/raw_failure_evidence.py:28]: Preserved. Lifecycle sampling is driven by the complete closed evidence vocabulary and support pairs, so every typed kind remains inspectable.

File: polylogue/pipeline/services/ingest_worker.py [hunk 9]
```diff
@@ -250,6 +251,7 @@ def _record_result(
             retryable=(disposition.retryable if disposition is not None else None),
             evidence_ref=(disposition.evidence_ref if disposition is not None else None),
             remediation=(disposition.remediation if disposition is not None else None),
+            diagnostic=(disposition.diagnostic if disposition is not None else None),
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/pipeline/services/ingest_worker.py:254]: Preserved and extended. The worker disposition fields now reach the ordinary batch source persistence boundary, with the current commit exercising diagnostic and typed fields through a real repository.

File: polylogue/sources/live/batch.py [hunk 10]
```diff
@@ -2806,6 +2832,15 @@ def _ingest_full_records_archive(
...
+                        if provider is Provider.UNKNOWN and _is_json_stream_decode_error(exc):
+                            archive.record_raw_failure_evidence(
+                                source_raw_id,
+                                provider=provider,
+                                source_path=record.source_path,
+                                source_index=record.source_index or 0,
+                                acquired_at_ms=acquired_at_ms,
+                                kind=RawFailureEvidenceKind.TERMINAL_UNKNOWN_JSON_DECODE,
+                            )
                         archive.mark_raw_parse_failed(source_raw_id, provider=provider, error=exc)
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/sources/live/batch.py:2875]: Preserved. Terminal unknown-input outcomes are treated as handled after durable evidence is written, so only deferred outcomes remain retryable.

File: polylogue/sources/decoder_json.py [hunk 11]
```diff
@@ -185,6 +205,9 @@ def _iter_jsonl_stream(
         )
         yield from records
 
+    if fail_on_decode_error and error_count:
+        raise JsonlDecodeError(path_name, line_number=line_number, cause=ValueError("malformed JSONL record"))
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/sources/decoder_json.py]: Preserved. The strict JSONL decoder retains the final physical record error count before selecting terminal unknown decode evidence.

File: polylogue/storage/repair.py [hunk 12]
```diff
@@ -3928,19 +3946,15 @@ def _raw_materialization_candidate_ids(
...
-                  -- is what keeps this retry-eligible even after the guard's
-                  -- own wording changes again.
-                  r.parse_error LIKE '{_MEMBERSHIP_REPLAY_CONFLICT_ERROR_PREFIX}%'
+                OR EXISTS (
+                  SELECT 1
+                  FROM raw_artifacts AS retry_evidence
+                  WHERE 1 = 1
+                    {_raw_artifact_coordinate_predicate(artifact_alias="retry_evidence", raw_alias="r")}
+                    AND retry_evidence.artifact_kind IN ({", ".join("?" for _ in RAW_FAILURE_DEFERRED_EVIDENCE_KINDS)})
+                    AND retry_evidence.support_status = ?
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/storage/repair.py]: Addressed. Successful async batch persistence now supersedes exact deferred CAS authority in the same source transaction, preventing stale replay authority after a resolved attempt.

[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/storage/repair.py]: Preserved. Retained-raw replay authorization remains limited to frontier conflicts; hot partial captures remain deferred until a complete observation.

File: polylogue/storage/sqlite/archive_tiers/revision_governance.py [hunk 13]
```diff
@@ -3066,16 +3083,77 @@ def record_raw_failure_evidence(
...
+        source_index=int(source_index or 0),
+        acquired_at_ms=int(time.time() * 1000),
+        kind=RawFailureEvidenceKind.TERMINAL_SUPERSEDED_DEFERRED_CAS_FRONTIER,
+    )
+
+
 def mark_raw_parse_succeeded(store: RawRevisionGovernanceHost, raw_id: str, *, provider: Provider) -> None:
     """Finalize one retained raw payload after every derived session commits."""
+    _supersede_deferred_cas_evidence(store, raw_id, provider=provider)
     finalize_raw_parse_state(store, raw_id, state=_raw_parse_success_state(provider))
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/storage/sqlite/archive_tiers/revision_governance.py]: Addressed in 28b99af28. The async batch success and skip loops now invoke deferred CAS supersession before raw-state persistence, and the real source persistence regression covers the stale carrier.

File: polylogue/sources/decoder_json.py [hunk 14]
```diff
@@ -177,14 +197,18 @@ def _iter_jsonl_stream(
...
             logger_obj,
             pending,
             is_last=True,
             path_name=path_name,
         )
+        error_count += new_errors
         yield from records
 
+    if fail_on_decode_error and error_count:
+        raise JsonlDecodeError(path_name, line_number=line_number, cause=ValueError("malformed JSONL record"))
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/sources/decoder_json.py]: Preserved. JSONL diagnostics report the physical offending line rather than the final stream counter.

File: polylogue/storage/sqlite/archive_tiers/revision_governance.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/storage/sqlite/archive_tiers/revision_governance.py]: Preserved. CAS supersession and parse-state mutation share the source transaction, including queued batch flush behavior.

File: polylogue/storage/sqlite/migrations/source/030_raw_failure_coordinate_carriers.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/storage/sqlite/migrations/source/030_raw_failure_coordinate_carriers.sql:24]: Preserved. Lifecycle recency and sampling use only valid closed raw-failure carriers, so ordinary artifact recency cannot mask the typed carrier.

File: polylogue/pipeline/services/ingest_batch/_core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/pipeline/services/ingest_batch/_core.py:1188]: Preserved and extended. The ordinary batch route writes the full typed disposition and exact coordinate through the source transaction.

File: polylogue/pipeline/services/ingest_batch/_core.py [hunk 15]
```diff
@@ -2266,16 +2272,31 @@ def _failed_raw_state_update(
...
+def _raw_failure_evidence_kind(outcome: _RawIngestOutcome | None) -> RawFailureEvidenceKind | None:
+    """Map terminal worker input outcomes to closed source-tier carriers."""
+    if outcome is None:
+        return None
+    try:
+        outcome_code = IngestOutcome.from_string(outcome.outcome_code)
+    except ValueError:
+        return None
+    return {
+        IngestOutcome.CORRUPT_INPUT: RawFailureEvidenceKind.TERMINAL_CORRUPT_INPUT,
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/pipeline/services/ingest_batch/_core.py:2295]: Addressed in 28b99af28. Validation-failed rows are explainable only by terminal corrupt-input or unknown JSON decode carriers, with zero-length and decode-failure persistence tests plus a negative unsupported-terminal test.

File: polylogue/pipeline/services/ingest_batch/_core.py [hunk 16]
```diff
@@ -2266,16 +2272,31 @@ def _failed_raw_state_update(
...
+        IngestOutcome.UNSUPPORTED_SHAPE: RawFailureEvidenceKind.TERMINAL_UNSUPPORTED_SHAPE,
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/pipeline/services/ingest_batch/_core.py:2296]: Addressed in 28b99af28. Scheduled parse selection excludes exact-coordinate valid terminal carriers while force-reparse still selects them; malformed support and coordinate negatives remain retryable.

File: polylogue/sources/live/batch.py [hunk 17]
```diff
@@ -2518,23 +2550,43 @@ def _ingest_full_records_archive(
...
+                                        fail_on_decode_error=provider is Provider.UNKNOWN,
+                                    )
+                                )
                         else:
-                            payloads = list(_iter_json_stream(BytesIO(payload), source_name))
+                            payloads = list(
+                                _iter_json_stream(
+                                    BytesIO(payload),
+                                    source_name,
+                                    fail_on_decode_error=provider is Provider.UNKNOWN,
```
[the remark this answers is not in the document]
Comment (Sinity) [reply polylogue/sources/live/batch.py:2587]: Addressed in 28b99af28. Both prefetch worker branches pass strict decoding for Provider.UNKNOWN, and the real mixed JSONL watcher route proves no partial session is cached or materialized.

**Comment by Sinity** [comment]
@circleci rerun
